src/main/scala/org/broadinstitute/hail/expr/AST.scala

```diff
@@ -1240,6 +1268,37 @@ case class ApplyMethod(posn: Position, lhs: AST, method: String, args: Array[AST
           }
       }
 
+    case (returnType, "flatMap", Array(Lambda(_, param, body))) =>
+      val localIdx = ec.a.length
+      val localA = ec.a
+      localA += null
+      val bodyFn = body.eval(ec.copy(st = ec.st + (param -> (localIdx, returnType))))
+
+      (returnType: @unchecked) match {
+        case TArray(_) =>
+          AST.evalCompose[IndexedSeq[_]](ec, lhs) { is =>
+            is.flatMap { elt =>
+              localA(localIdx) = elt
+              bodyFn().asInstanceOf[TraversableOnce[_]]
+            }
+          }
+        case TSet(_) =>
+          AST.evalCompose[Set[_]](ec, lhs) { s =>
+            s.flatMap { elt =>
+              localA(localIdx) = elt
+              bodyFn().asInstanceOf[TraversableOnce[_]]
+            }
+          }
+      }
+
+    case (returnType, "flatten", Array()) =>
+      (returnType: @unchecked) match {
+        case TArray(_) =>
+          AST.evalCompose[IndexedSeq[_]](ec, lhs) { is => is.asInstanceOf[IndexedSeq[IndexedSeq[_]]].flatten }
```

Comment:
The nested `IndexedSeq`s might be null in which case the flatten will fail with a NullPointerException.

Clearly, we need tests for these.